nussl/datasets/hooks.py

```diff
@@ -417,6 +423,125 @@ def process_item(self, item):
+                 make_mix: bool = False,
+                 threshold_db: float = -60.0,
+                 segment_dur: float = 4.0,
+                 hop_ratio: float = 0.5,
+                 verbose: bool = False,
+                 padding_mode: int = 0,
+                 **kwargs):
+        assert sample_rate is not None, f"Please provide a sample rate for the dataset."
+        self.threshold_db = threshold_db
+        self.segment_dur = segment_dur
+        self.hop_ratio = hop_ratio
+        self.sample_rate = sample_rate
+        self.salient_src = salient_src
+        self.folder = folder
+        self.verbose = verbose
+        self.padding_mode = padding_mode
+
+        self.mix_src = MixSourceFolder(
+            folder, mix_folder, source_folders, ext, make_mix, **kwargs
+        )
+        self.song_metadata = self._populate_metadata()
+        self.offset = 0.0
+
+
+        super(SalientExcerptMixSourceFolder, self).__init__(
+            self._get_mix,
+            len(self.song_metadata),
+            sample_rate=sample_rate,
+            **kwargs
+        )
+
+    def _populate_metadata(self):
+        """Makes a metadata dict of salient start times for each source."""
+
+        if self.verbose:
+            # TODO: logging.info()
+            logging.info(f'Preprocessing data at {self.folder}.')
+
+        # TODO: Balanced dataset!
+        items = self.mix_src.get_items(self.folder)
+        metadata = []
+        for item in tqdm.tqdm(items, disable=not self.verbose):
+            mix, sources = self.mix_src.get_mix_and_sources(item)
+            target_src = sources[self.salient_src]
+            target_audio = target_src.audio_data
+            salient_starts = utils.find_salient_starts(target_audio,
+                                                       self.segment_dur,
+                                                       self.hop_ratio,
+                                                       self.sample_rate,
+                                                       self.threshold_db)
+            for start in salient_starts:
+                metadata.append({
+                    'mixsrc_item': item,
+                    'start': start
+                })
+        return metadata
+
+    def _get_mix(self, placeholder, item):
+        """OnTheFly closure. Gets called for every iteration to build a batch. A placeholder param is necessary since
+        OnTheFly will pass itself as well as the item to this function"""
+        item = self.song_metadata[item]
+        mixsrc_item = item['mixsrc_item']
+        # Note that the onset needs to be passed to the AudioSignal class as the kwarg offset.
+        onset = item['start']
+        salient_excerpt = self.mix_src.process_item(mixsrc_item, offset=onset/self.sample_rate,
+                                                    duration=self.segment_dur)
+        return self._pad_output(salient_excerpt)
+
+    def _pad_output(self, excerpt):
```

Comment:
Okay, I'm thinking that it's best not to support padding at all. It kind of makes the logic a complicated mess, so my feeling is that we should stick with just truncating. There are some edge cases, like "what if all of the audio files are too short?" and I think in that case, we should steer the user to another object (MixSourceFolder), or to change the parameters of this object. If there's a good use case for padding in the future then we can think about adding it. 

For now, just make sure the code supports truncating the last excerpt and maybe open an issue to talk about padding the last window.

Reply:
So this means that in `_get_mix()` we always assume the salient_excerpt is the correct length. The easiest way to accomplish this is to add a little bit of extra logic in the for loop at line 513 to just check if the window is long enough: `if start + self.segment_dur <= mix.duration: # keep`